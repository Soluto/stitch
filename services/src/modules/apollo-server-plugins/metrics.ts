import {
  ApolloServerPlugin,
  GraphQLRequestContextWillSendResponse,
  GraphQLFieldResolverParams,
} from 'apollo-server-plugin-base';
import * as promClient from 'prom-client';
import {
  knownApolloDirectives,
  requestDurationPromHistogramBuckets,
  resolverDurationPromHistogramBuckets,
} from '../config';

let requestDurationHistogram: promClient.Histogram<string> | undefined;
let resolverDurationHistogram: promClient.Histogram<string> | undefined;
let requestParsingErrorCounter: promClient.Counter<string> | undefined;
let requestValidationErrorCounter: promClient.Counter<string> | undefined;

export function initializeMetrics(pClient: typeof promClient) {
  requestDurationHistogram = new pClient.Histogram({
    name: 'graphql_request_duration_seconds',
    help: 'request duration in seconds',
    labelNames: ['status', 'operationName'],
    buckets: requestDurationPromHistogramBuckets,
  });

  resolverDurationHistogram = new pClient.Histogram({
    name: 'graphql_resolver_duration_seconds',
    help: 'resolver duration in seconds',
    labelNames: ['parentType', 'fieldName', 'status'],
    buckets: resolverDurationPromHistogramBuckets,
  });

  requestParsingErrorCounter = new pClient.Counter({
    name: 'graphql_request_parsing_errors_count',
    help: 'request query parsing errors',
  });

  requestValidationErrorCounter = new pClient.Counter({
    name: 'graphql_request_validation_errors_count',
    help: 'request query validation errors',
  });
}

export function createMetricsPlugin(): ApolloServerPlugin {
  return {
    async requestDidStart() {
      const endResponseTimer = requestDurationHistogram?.startTimer();
      return {
        async parsingDidStart() {
          return async (err?: Error) => {
            if (err) {
              requestParsingErrorCounter?.inc(1);
            }
          };
        },
        async validationDidStart() {
          return async (err?: ReadonlyArray<Error>) => {
            if (err && err?.length > 0) {
              requestValidationErrorCounter?.inc(1);
            }
          };
        },
        async willSendResponse(willSendResponseContext: GraphQLRequestContextWillSendResponse<unknown>) {
          const {
            operationName,
            response: { errors },
          } = willSendResponseContext;
          endResponseTimer?.({ operationName: operationName ?? 'N/A', status: errors ? 'error' : 'success' });
        },
        executionDidStart: async () => ({
          willResolveField(
            fieldResolverParams: GraphQLFieldResolverParams<unknown, unknown, Record<string, unknown>>
          ): ((error: Error | null, result?: unknown) => void) | void {
            const {
              source,
              info: { fieldName, parentType },
            } = fieldResolverParams;

            const parentObject = source as Record<string, unknown>;
            const intuitiveValue = parentObject?.[fieldName];

            const fieldDirectives = parentType.getFields()[fieldName].astNode?.directives;
            const hasCustomDirectives = fieldDirectives?.some(d => !knownApolloDirectives.has(d.name.value));

            if (intuitiveValue && !hasCustomDirectives) {
              // TODO: This is patch that should remove all fields which are resolved by default resolver.
              // Unfortunately, this patch will remove also fields with custom resolver that replaces source object property with name like a field
              return;
            }

            const endResolverTimer = resolverDurationHistogram?.startTimer({ parentType: parentType.name, fieldName });
            return (error: Error | null) => {
              endResolverTimer?.({ status: error ? 'error' : 'success' });
            };
          },
        }),
      };
    },
  };
}
