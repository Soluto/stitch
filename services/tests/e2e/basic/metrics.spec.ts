import { print } from 'graphql';
import { gql } from 'apollo-server-core';
import fetch from 'node-fetch';
import { GraphQLClient } from 'graphql-request';
import PrometheusMetricsSerializer from '../../utils/prometheus-metrics-serializer';
import { getToken } from '../../helpers/get-token';
import { RegistryMutationResponse, updateSchemasMutation } from '../../helpers/registry-request-builder';
import { updateGatewaySchema } from '../../helpers/utility';
import { schema1, schema2 } from './metrics.schema';

const gatewayBaseUrl = 'http://localhost:8080';
const registryBaseUrl = 'http://localhost:8090';

const gatewayClient = new GraphQLClient(`${gatewayBaseUrl}/graphql`);
const registryClient = new GraphQLClient(`${registryBaseUrl}/graphql`);

const gatewayMetricsEndpoint = `${gatewayBaseUrl}/metrics`;
const registryMetricsEndpoint = `${registryBaseUrl}/metrics`;

async function getMetrics(endpoint: string) {
  const response = await fetch(endpoint);
  expect(response.ok).toBeTruthy();
  const body = await response.text();
  return body.split('\n');
}

describe('Metrics', () => {
  const query = print(gql`
    query GetFoo {
      m_foo {
        bar
        baz
        # taz TODO: Fix registry to accept Apollo Federation directives
      }
    }
  `);

  beforeAll(async () => {
    expect.addSnapshotSerializer(PrometheusMetricsSerializer);

    const accessToken = await getToken();
    gatewayClient.setHeader('Authorization', `Bearer ${accessToken}`);

    const registryAccessToken = await getToken({ scope: 'stitch-registry' });
    registryClient.setHeader('Authorization', `Bearer ${registryAccessToken}`);
  });

  test('Setup schema', async () => {
    const updateSchemaResponse = await registryClient.request<RegistryMutationResponse>(updateSchemasMutation, {
      schema: schema1,
    });
    expect(updateSchemaResponse.result.success).toBeTruthy();

    const resp = await updateGatewaySchema('http://localhost:8080');
    expect(resp.status).toEqual(200);

    const response = await gatewayClient.request(query);
    expect(response).toBeDefined();
  });

  test('Check metrics', async () => {
    const metrics = await getMetrics(gatewayMetricsEndpoint);
    const graphqlMetrics = metrics.filter(m => m.includes('graphql_'));
    expect(graphqlMetrics).toMatchSnapshot();
  });

  test('Change schema', async () => {
    const updateSchemaResponse = await registryClient.request<RegistryMutationResponse>(updateSchemasMutation, {
      schema: schema2,
    });
    expect(updateSchemaResponse.result.success).toBeTruthy();

    const resp = await updateGatewaySchema('http://localhost:8080');
    expect(resp.status).toEqual(200);
  });

  test('Check metrics again', async () => {
    const metrics = await getMetrics(gatewayMetricsEndpoint);
    expect(metrics).toContainEqual(
      'graphql_resolver_duration_seconds_count{parentType="MetricsFoo",fieldName="baz",status="success"} 1'
    );
  });

  test('Parsing errors metric', async () => {
    await expect(gatewayClient.request('wrong query')).rejects.toMatchSnapshot();

    const metrics = await getMetrics(gatewayMetricsEndpoint);
    expect(metrics).toContainEqual('graphql_request_parsing_errors_count 1');
  });

  test('Validation errors metric', async () => {
    const query = print(gql`
      query {
        wrong
      }
    `);
    await expect(gatewayClient.request(query)).rejects.toMatchSnapshot();

    const metrics = await getMetrics(gatewayMetricsEndpoint);
    expect(metrics).toContainEqual('graphql_request_validation_errors_count 1');
  });

  test('Check resolver histogram metrics buckets', async () => {
    const metrics = await getMetrics(gatewayMetricsEndpoint);
    const bucketMetrics = metrics.filter(
      m => m.startsWith('graphql_resolver_duration_seconds_bucket') && m.includes('MetricsFoo')
    );
    const bucketRegex = /le="(.+)",parentType=/;
    const buckets = bucketMetrics.map(m => m.match(bucketRegex)?.[1]);
    expect(buckets).toHaveLength(4);
    expect(buckets).toMatchSnapshot();
  });

  test('Check request histogram metrics buckets', async () => {
    const metrics = await getMetrics(gatewayMetricsEndpoint);
    const bucketMetrics = metrics.filter(
      m => m.startsWith('graphql_request_duration_seconds_bucket') && m.includes('GetFoo')
    );
    const bucketRegex = /le="(.+)",operationName=/;
    const buckets = bucketMetrics.map(m => m.match(bucketRegex)?.[1]);
    expect(buckets).toHaveLength(5);
    expect(buckets).toMatchSnapshot();
  });

  test('Registry metrics', async () => {
    const metrics = await getMetrics(registryMetricsEndpoint);
    const sampleMetric = metrics.find(m => m.startsWith('nodejs_version_info'));
    expect(sampleMetric).toBeDefined();
    const majorVersion = /major="(\d+)"/g.exec(sampleMetric!);
    expect(majorVersion?.[1]).toEqual('16');
  });
});
