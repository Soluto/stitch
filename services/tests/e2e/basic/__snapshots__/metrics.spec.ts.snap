// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Metrics Check metrics 1`] = `
graphql_resolver_duration_seconds_bucket{le="X",parentType: Query,fieldName: m_foo,status: success} Y
graphql_resolver_duration_seconds_sum{parentType: Query,fieldName: m_foo,status: success} Y
graphql_resolver_duration_seconds_count{parentType: Query,fieldName: m_foo,status: success} Y
graphql_resolver_duration_seconds_bucket{le="X",parentType: MetricsFoo,fieldName: baz,status: success} Y
graphql_resolver_duration_seconds_sum{parentType: MetricsFoo,fieldName: baz,status: success} Y
graphql_resolver_duration_seconds_count{parentType: MetricsFoo,fieldName: baz,status: success} Y
`;

exports[`Metrics Check request histogram metrics buckets 1`] = `
Array [
  "0.02",
  "0.1",
  "0.5",
  "2",
  "+Inf",
]
`;

exports[`Metrics Check resolver histogram metrics buckets 1`] = `
Array [
  "0.02",
  "0.1",
  "0.5",
  "+Inf",
]
`;

exports[`Metrics Parsing errors metric 1`] = `[Error: Syntax Error: Unexpected Name "wrong".: {"response":{"errors":[{"message":"Syntax Error: Unexpected Name \\"wrong\\".","extensions":{"code":"GRAPHQL_PARSE_FAILED"}}],"status":400},"request":{"query":"wrong query"}}]`;

exports[`Metrics Validation errors metric 1`] = `[Error: Cannot query field "wrong" on type "Query".: {"response":{"errors":[{"message":"Cannot query field \\"wrong\\" on type \\"Query\\".","extensions":{"code":"GRAPHQL_VALIDATION_FAILED"}}],"status":400},"request":{"query":"{\\n  wrong\\n}\\n"}}]`;
